nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+import org.apache.commons.io.FilenameUtils;
+import org.nuxeo.common.utils.Path;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.impl.blob.ZipEntryBlob;
+import org.nuxeo.ecm.platform.filemanager.api.FileImporterContext;
+import org.nuxeo.ecm.platform.filemanager.api.FileManager;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+/** @since 2021.0 */
+public class CompoundDocumentServiceImpl extends DefaultComponent implements CompoundDocumentService {
+
+    protected static final String XP_IGNORED_FILES = "ignoredFiles";
+
+    @Override
+    public DocumentModel createCompoundDocument(CoreSession session, String parent, Blob archive) {
+        String compoundDocName = FilenameUtils.removeExtension(archive.getFilename());
+        DocumentModel compoundDoc = session.createDocumentModel(parent, compoundDocName, COMPOUND_DOCTYPE);
+        compoundDoc = session.createDocument(compoundDoc);
+        createCompoundDocument(compoundDoc, archive);
+        return compoundDoc;
+    }
+
+    protected void createCompoundDocument(DocumentModel compoundDoc, Blob archive) {
+        try (ZipFile zip = new ZipFile(archive.getFile())) {
+            FileManager fileManager = Framework.getService(FileManager.class);
+            zip.stream().filter(this::isAllowedEntry).sorted(Comparator.comparing(ZipEntry::getName)).forEach(entry -> {
+                Path entryPath = new Path(entry.getName());
+                String parentPath = compoundDoc.getPath().append(entryPath.removeLastSegments(1)).toString();
+                if (entry.isDirectory()) {
+                    CoreSession session = compoundDoc.getCoreSession();
+                    var doc = session.createDocumentModel(parentPath, entryPath.lastSegment(), COMPOUND_FOLDER_DOCTYPE);
+                    session.createDocument(doc);
+                } else {
+                    ZipEntryBlob blob = new ZipEntryBlob(zip, entry);
+                    blob.setFilename(entryPath.lastSegment());
+                    FileImporterContext ctx = FileImporterContext.builder(compoundDoc.getCoreSession(), blob,
+                            parentPath).build();
+                    try {
+                        fileManager.createOrUpdateDocument(ctx);
+                    } catch (IOException e) {
+                        String message = String.format(
+                                "Failed to create CompoundDocument: %s. something went wrong with archive: %s at entry: %s.",
+                                compoundDoc.getName(), archive.getFilename(), entry.getName());
+                        throw new NuxeoException(message, e);
+                    }
+                }
+            });
```

Comment:
I would put all of this in a specific method to ease readability, avoid too many indents, and nested try-catch blocks.